### ISSUE TITLE
GCE upgrade script is dangerously unreliable in some situations

### ISSUE THREAD

**Issue by paralin** [kubernetes/kubernetes#23004]
This happens when I tried to do an upgrade:

```
Your active configuration is: [default]

Project: faceit-ltd
Zone: us-central1-b
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.2.0-beta.1/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
It looks as if you don't have a compiled kubectl binary

If you are running from a clone of the git repo, please run
'./build/run.sh hack/build-cross.sh'. Note that this requires having
Docker installed.

If you are running from a binary release tarball, something is wrong. 
Look at http://kubernetes.io/ for information on how to contact the 
development team for help.
It looks as if you don't have a compiled kubectl binary

If you are running from a clone of the git repo, please run
'./build/run.sh hack/build-cross.sh'. Note that this requires having
Docker installed.

If you are running from a binary release tarball, something is wrong. 
Look at http://kubernetes.io/ for information on how to contact the 
development team for help.
It looks as if you don't have a compiled kubectl binary

If you are running from a clone of the git repo, please run
'./build/run.sh hack/build-cross.sh'. Note that this requires having
Docker installed.

If you are running from a binary release tarball, something is wrong. 
Look at http://kubernetes.io/ for information on how to contact the 
development team for help.
```

I actually did have kubectl compiling in the background and didn't realize it hadn't finished when I did the upgrade. But this resulted in a broken cluster. Initially it couldn't find the master, as it has deleted it. I did some frantic hacking.

```
Your active configuration is: [default]

Project: faceit-ltd
Zone: us-central1-b
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.2.0-beta.1/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
ERROR: (gcloud.compute.instances.create) Could not read metadata key [cluster-name] from file [/tmp/kubernetes.8KVywX/cluster-name.txt]: No such file or directory
```

Manually overrode this with "kubernetes" and skipped the existing master checks, and now I have a working master again.

This is pretty bad, though. Obviously the most simple fix is to just check for kubectl before doing upgrade, sure. But I think we should also introduce a "create-master" script in the situation that the master was deleted and needs to be re-made.

Also master should be put under a autoscaling group.

Edit: I just tried to do the upgrade again and now this shit happens:

```
Your active configuration is: [default]

Project: faceit-ltd
Zone: us-central1-b
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.2.0-beta.1/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
Using master: kubernetes-master (external IP: 130.211.145.169)
Deleted [https://www.googleapis.com/compute/v1/projects/faceit-ltd/zones/us-central1-b/instances/kubernetes-master].
ERROR: (gcloud.compute.instances.create) Could not read metadata key [cluster-name] from file [/tmp/kubernetes.8yL2f8/cluster-name.txt]: No such file or directory
```

and now I don't have a master again...

**Comment by mbrevoort** [comment]
My upgrade just failed in a similar fashion:

```
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.2.2/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
Using master: kubernetes-master (external IP: 130.211.137.128)
Deleted [https://www.googleapis.com/compute/v1/projects/beepboop-1053/zones/us-central1-b/instances/kubernetes-master].
ERROR: (gcloud.compute.instances.create) Could not read metadata key [cluster-name] from file [/var/folders/hr/f3mv1fm51ds751c9q0bcwmjc0000gn/T/kubernetes.XXXXXX.OIi95GCC/cluster-name.txt]: No such file or directory
```

I'm now stuck with a cluster with no master node. 😭

**Comment by janetkuo** [comment]
Upgrading master from 1.2.4 to 1.3.0-beta.2 failed :( 
My new master never respond to API requests...

Running script from HEAD:  
```console
➜ kubernetes(master)$ cluster/gce/upgrade.sh -M v1.3.0-beta.2
Your active configuration is: [default]

Project: kuber-3
Zone: us-central1-b
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 38796  100 38796    0     0   130k      0 --:--:-- --:--:-- --:--:--  130k
== Upgrading master to 'https://storage.googleapis.com/kubernetes-release/release/v1.3.0-beta.2/kubernetes-server-linux-amd64.tar.gz'. Do not interrupt, deleting master instance. ==
Using master: kubernetes-master (external IP: 146.148.35.19)
Enter passphrase for key '/usr/local/google/home/xxxx/.ssh/google_compute_engine': 
Deleted [https://www.googleapis.com/compute/v1/projects/kuber-3/zones/us-central1-b/instances/kubernetes-master].
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/kuber-3/zones/us-central1-b/instances/kubernetes-master].
NAME               ZONE           MACHINE_TYPE   PREEMPTIBLE  INTERNAL_IP  EXTERNAL_IP    STATUS
kubernetes-master  us-central1-b  n1-standard-1               10.240.0.2   146.148.35.19  RUNNING
== Waiting for new master to respond to API requests ==
...................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
```

**Comment by janetkuo** [comment]
cc @wojtek-t @piosz

**Comment by piosz** [comment]
@paralin @mbrevoort your issue was fixed in #27763

@janetkuo copy/pasting your problem to and closing in favor of #27764
